### PR TITLE
🐰 Add a basic fuzz target

### PR DESCRIPTION
**Pull request by CosmicHorrorDev**
This adds a fuzz target to ensure that parsing doesn't panic

If you wanted to get a bit fancier and had some code for rendering the document back to a string then you could fuzz the full parse -> render -> parse and then check if the tables from both parses are equivalent to ensure semantic equivalence

### CODE REVIEW

**Comment by CosmicHorrorDev** [comment]
Thinking on it more I think adding another target for comparing the parsed value to `toml-edit` would be better than a parse -> render roundtrip

**Comment by zeenix** [comment]
> This adds a fuzz target to ensure that parsing doesn't panic

Nice! Thanks so much for contributing this.

> Thinking on it more I think adding another target for comparing the parsed value to toml-edit would be better than a parse -> render roundtrip

Yeah, we could definitely have `toml-edit` as a dev dependency and do this.

> Edit: eh that would require us following strict compliance with the spec which I now see listed as a non-goal

I made that a non-goal only because I was anticipating maintaining and developing this most alone. If you're going to continue contributing, we can definitely make it a goal, as long as it doesn't conflict with the other goals, especially to keep the crate simple and small.

File: fuzz/fuzz_targets/parse_no_panic.rs [hunk 1]
```diff
@@ -1,7 +1,8 @@
 #![no_main]
 
 use libfuzzer_sys::fuzz_target;
+use tomling::parse;
 
-fuzz_target!(|data: &[u8]| {
-    // fuzzed code goes here
+fuzz_target!(|input: &str| {
+    let _ = parse(input);
```
Comment (zeenix) [inline fuzz/fuzz_targets/parse_no_panic.rs:7]: Looks like the second commit is a fixup of the first commit? Please squash them if so.

**Review by zeenix — CHANGES_REQUESTED** [review]
As I wrote, really happy about this contribution. Two other comments:

* It would be great to have this run as part of the CI, like I'm doing for zvariant subcrate of zbus [here](https://github.com/dbus2/zbus/blob/main/.github/workflows/rust.yml#L208). I'd prefer it if you could already do that as part of this PR (obviously this would mean we must fix all issues we find through fuzzing before merging this PR but I'm ok with that) but I'm also fine with doing that in a follow-up PR. Up to you.
* As per [the contribution guide](https://github.com/zeenix/tomling/blob/main/CONTRIBUTING.md#submitting-pull-requests), could you please add emoji prefixes to your commits, either by setting up [`gimoji`](https://github.com/zeenix/gimoji) for your local repo (this ensure you don't forget to add one) or by copy&pasting them from [here](https://gitmoji.dev/)? I'm not really strict about it with flyby contributions but seems like you'd be contributing more than 1-2 commits so I'd appreciate it if you could follow this practice. :pray:

**Comment by CosmicHorrorDev** [comment]
I can address all your comments after I'm off work

**Comment by zeenix** [comment]
> I can address all your comments after I'm off work

No! Now! now! now! :laughing: Of course! take your time, no hurry!

**Comment by CosmicHorrorDev** [comment]
Squashed and fuzzy emoji added

> If you're going to continue contributing, we can definitely make it a goal, as long as it doesn't conflict with the other goals, especially to keep the crate simple and small.

I'd be willing to help maintain this crate. I've still got more PRs planned anyways :grin:

> * It would be great to have this run as part of the CI, like I'm doing for zvariant subcrate of zbus [here](https://github.com/dbus2/zbus/blob/main/.github/workflows/rust.yml#L208). I'd prefer it if you could already do that as part of this PR (obviously this would mean we must fix all issues we find through fuzzing before merging this PR but I'm ok with that) but I'm also fine with doing that in a follow-up PR. Up to you.

Sure I can add that to this PR :+1:

I didn't find any crashes beyond #9 after working around #9 and #8 to fuzz for around a 100 million iterations

**Comment by zeenix** [comment]
> Squashed and fuzzy emoji added

Thanks but the emoji needs to be one of the curated ones. That's why I recommend using one of the two methods. The point of the emoji is not just so it looks cute and nice (that's a bonus :smile:) but rather making it easy for highly-visual folks like me to read and search through the git log and release notes created out of the git log.

Having said that, we don't exactly seem to have a specific emoji for fuzzing and :rabbit: is a great choice. [I'll add it](https://github.com/zeenix/gimoji/issues/190) to the gimoji list.

> > If you're going to continue contributing, we can definitely make it a goal, as long as it doesn't conflict with the other goals, especially to keep the crate simple and small.
> 
> I'd be willing to help maintain this crate. I've still got more PRs planned anyways 😁

That's very great to hear. In that case, I'm more than happy to remove strict compliance with the spec from our non-goals.

> Sure I can add that to this PR 👍

Nice!
 
> I didn't find any crashes beyond #9 after working around #9 and #8 to fuzz for around a 100 million iterations

Cool. I also didn't see many `unwraps` used in the parser so that makes sense. I'll try to fix #9 today along with un-ignoring some of the related tests.

**Comment by CosmicHorrorDev** [comment]
Should be ready for another review :)

**Review by zeenix — APPROVED** [review]
LGTM. Perhaps we should solve #8 before merging this?

**Comment by CosmicHorrorDev** [comment]
I think it's fine merging this already. #8 can cause the fuzzing to slow down a lot, but it won't cause failures which I think is fine

**Pull request merged**